src/global_distrib/mod_global_distrib.erl

```diff
@@ -29,34 +29,52 @@
 -export([deps/2, start/2, stop/1, config_spec/0]).
 -export([find_metadata/2, get_metadata/3, remove_metadata/2, put_metadata/3]).
 -export([maybe_reroute/1]).
--export([process_endpoints/1, process_bounce/1]).
+-export([process_opts/1, process_endpoint/1]).
 
 -ignore_xref([maybe_reroute/1, remove_metadata/2]).
 
 %%--------------------------------------------------------------------
 %% gen_mod API
-%% See "gen_mod logic" block below in this file
 %%--------------------------------------------------------------------
 
--spec deps(Host :: jid:server(), Opts :: proplists:proplist()) -> gen_mod_deps:deps().
-deps(Host, Opts) ->
-    mod_global_distrib_utils:deps(?MODULE, Host, Opts, fun deps/1).
+-spec deps(mongooseim:host_type(), gen_mod:module_opts()) -> gen_mod_deps:deps().
+deps(HostType, Opts = #{global_host := HostType, bounce := BounceOpts}) ->
+    %% Start each required module with the same opts for simplicity
+    [{Mod, Opts, hard} || Mod <- dep_modules() ++ bounce_modules(BounceOpts)];
+deps(_HostType, #{}) ->
+    [].
 
--spec start(Host :: jid:lserver(), Opts :: proplists:proplist()) -> any().
-start(Host, Opts0) ->
-    Opts = [{message_ttl, 4} | Opts0],
-    mod_global_distrib_utils:start(?MODULE, Host, Opts, fun start/0).
+dep_modules() ->
+    [mod_global_distrib_utils, mod_global_distrib_mapping, mod_global_distrib_disco,
+     mod_global_distrib_receiver, mod_global_distrib_hosts_refresher].
 
--spec stop(Host :: jid:lserver()) -> any().
-stop(Host) ->
-    mod_global_distrib_utils:stop(?MODULE, Host, fun stop/0).
+bounce_modules(#{enabled := true}) -> [mod_global_distrib_bounce];
+bounce_modules(#{enabled := false}) -> [].
+
+-spec start(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start(HostType, #{global_host := HostType}) ->
+    mongoose_metrics:ensure_metric(global, ?GLOBAL_DISTRIB_DELIVERED_WITH_TTL, histogram),
+    mongoose_metrics:ensure_metric(global, ?GLOBAL_DISTRIB_STOP_TTL_ZERO, spiral),
+    ejabberd_hooks:add(hooks());
+start(_HostType, #{}) ->
```

Comment:
Will this ever match? Similarly in `deps`.

Reply:
Yes. It will match for all host types different than `global_host`, e.g. `local_host`. This module should be enabled at the top level, so it is always enabled for all hosts. However, the initialization and deps setup needs to be done only for `global_host`.

Reply:
Ah, of course, I see. I missed that `global_host` matches to the same `Host(Type)` as the module is started for. Makes sense, thanks for the explanation.